Refactor

balazs482 opened the pull request:
## @hegyhati kész a refactor!
   
- A Character most ősosztálya a Heronak (korábban Player) és az újonnan készült Monsternek
- A JSON (korábban JSONParser) már osztály (korábban névtér), hogy tartalmazhasson saját mapot  
- Függvények és változók pótolva vagy átnevezve   
- A konstruktorok átalakítva   
- Workflowk átalakítva   
- Az érvényét vesztett dokumentáció részek törölve   
- Fixálások érkeztek a parserre   
- Readme frissítve   
- Új teszt létrehozva   
  
A fájlnév változtatások git mv paranccsal készültek.

gboldi19 committed:
Created unified_test.yml as it doesnt let me push from local

gboldi19 committed:
Updating the deployment of the documentation

balazs482 committed:
Unhandled error fix.

Legit colon being the last printable character of the .json now throws error.

balazs482 committed:
Type-o fix.

Accidental bracket ending symbol added to the code before upload.

balazs482 committed:
: as last printable char error throw finalized

balazs482 committed:
Bug fix on map value recognition.

Had not deleted extra characters.

balazs482 committed:
Refactor to new main.cpp...

PR message will contain refactor changes.

Unnecessary inclusions removed.

Two structural changes:
1. JSON (previously JSONParser) is now class instead of namespace.
2. Character is now base class of both Hero (previously Player) and newly made Monster classes.

New scenario files created and test script modified in order to run workflows properly.

The expected output may require changes as a simulation bug has also been discovered and solved.

balazs482 committed:
GNU compiler now on newer version.

Default version could not find <filesystem>.

balazs482 committed:
Update expected_output.csv

balazs482 committed:
Update CMakeLists.txt

(Find and replace somehow missed this file.)

balazs482 committed:
Update README.md

balazs482 committed:
Parser now accepts some special characters within strings

balazs482 committed:
Merge branch 'refactor' of https://github.com/Teaching-projects/SZE-MOSZE-2020-Mortar_Kombat into refactor

balazs482 committed:
Parser now can return NULL.

All single data values are now available to use.

balazs482 committed:
Refactored to new .json format.

Added separate Hero and Monster type testing units, because the .json format differs.

NOTE: The level up process is still the same.

balazs482 committed:
Variable namechange.

camelCase implies XPToLevelup instead of XPtoLevelup.

balazs482 committed:
Indentation fixed

balazs482 committed:
Now using new tester file

balazs482 committed:
Now tests for null pointer

balazs482 committed:
Now really tests for null pointer

balazs482 committed:
Now uses NULL instead of nullptr.

NULL's void* type automatically converts to int, but nullptr's nullptr_t type results in a bad_anycast error.

balazs482 committed:
Now tests for NULL value

balazs482 committed:
Now tests for NULL value.

Missing brackets added.

balazs482 committed:
Now tests for null pointer.

"null" as a value is stored as a null pointer.

balazs482 committed:
Temporary inclusion removed.

balazs482 committed:
Newly required methods of levelup

balazs482 committed:
More logical fighting function

balazs482 committed:
Accidental inclusion removed

balazs482 committed:
Update expected_output.csv

...new level up method required changes.

balazs482 committed:
Accidental new line removed.

Manually created expected output had an extra line causing false positive diffs.

balazs482 committed:
New scenario test, workflows reworked.

balazs482 committed:
Update README.md

hegyhati commented:
Te sem tudsz aludni, latom :-D A PR nyitokomment csillagos otos. A tobbit majd csak este vagy esetleg kora delutan fogom csak tudni megnezni.

balazs482 committed:
Indentation fix

hegyhati commented on Hero.cpp:
```diff
@@ -0,0 +1,67 @@
+	JSON::ruleOutNegativeAnyFloat(characterAttributes["damage_bonus_per_level"]);
+	JSON::ruleOutNegativeAnyFloat(characterAttributes["cooldown_multiplier_per_level"]);
+	return Hero(
+        std::any_cast<std::string>(characterAttributes["name"]),
+        std::any_cast<float>(characterAttributes["base_health_points"]),
+        std::any_cast<float>(characterAttributes["base_damage"]),
+        std::any_cast<float>(characterAttributes["base_attack_cooldown"]),
+		std::any_cast<float>(characterAttributes["experience_per_level"]),
+		std::any_cast<float>(characterAttributes["health_point_bonus_per_level"]),
+		std::any_cast<float>(characterAttributes["damage_bonus_per_level"]),
+		std::any_cast<float>(characterAttributes["cooldown_multiplier_per_level"]));
```
it miert nem hasznalod a get-et? azert lett csinalva. ugyanez monsternel., hogy végre legyen virtuális futás, küldöm a selfie-t a montázshoz ￼

balazs482 replied:
Ez még a korai main.cpp hozománya, amiben nem volt a JSON-nak saját map tagváltozója, nem kellett getterezni.

balazs482 replied:
### @hegyhati kész vagyok a módosításokkal.   
   
### Emellett a documentation branchről lenne pár kérdés:  
#### 1  
A documentation ágat a refactor ágból indítottam. Már kész a documentation PR is. Az a PR is a masterbe való mergere van nyitva. Amíg ez nem megy masterre, addig azt Draft-ként hagyom. Csak szerettem volna, hogy elinduljon a review ott is.   
#### 2   
Az osztályoknak elég a public tagjaihoz írni doksit?  
#### 3   
Most úgy van beállítva a documentation deploying workflow, hogy PR nyitáskor, illetve nyitott PR branchére commitoláskor fusson. Erre mi lenne jobb megoldás? Master frissülésekor frissüljön? Mindig egy documentation nevű branchre pusholáskor készüljön el? Egyszerűen kapcsoljuk ki az idő nagy részében az actiont?  

#### Előre köszönöm a válaszokat!

hegyhati replied:
wtf was I talking about? Pedig nem is ittam. (amugy az egerem kozepso gombja neha onallo eletre kel, az meg beszurja a legutoljara kijelolt dolgot, szoval ez tortent valoszinuleg es nem vettem eszre. sry about that.)

A modositasokat mindjart nezem. 
1 ok
2 en legalabb a protectedhez megirnam meg, mert ha valaki szarmaztat belole, azt meg ugye  eleri. 
3 en azt csinalnam, hogy "master" es "documentation" branchekre menjen le. Az esetek 90%-ban magan a deployolasi folyamaton nem valtoztattok, szoval azt nem kell tesztelni, branchen lefuttatni. Az rossz is lenne, ha a brancheken levo felkesz dolgokkal frissulne. A master a lenyeg.A document branchre meg ugy futtatnam, hogy egy gh-pages-2-re deployoljon tesztnek. Es akkor ha magan a doxygenen hegesztetek valamit akkor azt ott csinalnam.

hegyhati reviewed (changes requested):
az any cast logika egy helyen eleg ha megvan, a get-ben, mindenhol mashol meg lehet azt hasznalni.
Amugy rendben levonek tunik.

balazs482 committed:
Character construction from file simplified.

Now JSON::RONAF (previously JSON::ruleOutNegativeAnyFloat) can be called inside the return of <Character subclass>::parse() functions. Also, these functions now store data in a JSON object, so that they can use the getter. This solution looks and performs better.

balazs482 committed:
Simplified JSON functions.

Now uses std::string::erase() and advanced for loops.

hegyhati reviewed (approved):
Rendben van, bar en ezt a RONAF-ot kiszedem a JSON-bol. Nem a JSON-nek fontos, hogy valami negativ-e vagy sem. Az exceptiont a Hero konstruktorba pakolnam, hogy negativ adat lett megadva, pedig ott nem szereti a negativ adatot.  De bemehet igy is. en approvoltam, szoljatok, ha vigyem be / modositottatok es akkor vigyem be.

balazs482 committed:
RONAF moved to Character header.

Rule Out Negative Any Float function moved from JSON class to Character header file (not inside the class, because it is only a helper function with a specific implication in the subclasses).

balazs482 commented:
@hegyhati 
Teljesen igaz, át is írtam. Ha minden jól megy, most lefut a workflow és mehet a merge.

gboldi19 commented:
@hegyhati éppen írtam a kommentet, amikor bement D: azt szerettem volna kérni, hogy ez a branch mehetne először és majd oldok conflictokat ott, de már mindegy, megcsináljuk itt

hegyhati commented:
:-D jol van, szoljatok, ha fel van oldva, es akkor viszem be.

gboldi19 committed:
Trying to fix upstream bug

hegyhati commented:
Hm... most akkor egy elcrashelo valtozatot vittetek be Ti a masterbe? not cool.
